doc/03_reference/testplan.rst

```diff
@@ -0,0 +1,109 @@
+.. _testplan:
+
+.. todo::
+
+  Add detail about security hardening verification.
+
+.. note::
+
+  This testplan is a work in progress still being implemented so this document may not match the implemented verification in the repository.
+
+Test Plan
+=========
+
+Goals
+-----
+
+* Verify compliance with all the RISC-V specifications Ibex supports.
+* Verify Ibex's security hardening features.
+* Ensure correct functionality is maintained across all possible behaviours of external interfaces (interrupts, memory responses, debug requests etc).
+* Hit all functional coverage points, described in :ref:`coverage-plan`.
+
+Testbench Architecture
+----------------------
+
+.. figure:: images/tb2.svg
+    :alt: Testbench Architecture
+
+    Architecture of the UVM testbench for Ibex core
+
+Ibex utilises a co-simulation checking approach described in detail in :ref:`cosim`.
+With the co-simulation system all instructions Ibex executes and all external events such as an interrupts or memory errors are fed to a golden model.
+The results of every instruction execution and every memory access are crossed checked against the golden model with any mismatches resulting in a test failure.
+The aim is to check all possible externally observable behaviours of ``ibex_top`` against the golden model.
+The golden model used is the `Spike RISC-V ISS <https://github.com/riscv-software-src/riscv-isa-sim>`_.
+
+The testbench uses UVM.
+It consists of 3 agents:
+
+Co-simulation Agent:
+  This has multiple monitors.
+  One monitors the RVFI interface which provides details of retired instructions.
+  The other monitors relate to fetched instructions and instruction memory errors, more details are provided in :ref:`cosim`.
+  Additionally it connects to the monitor of the Memory Interface Agent for the instruction and data side via analysis ports.
+  The monitored transactions are used by a scoreboard to provide information to the co-simulation system allowing it to step the golden model and check its execution and memory activity against Ibex's behaviour.
+
+Memory Interface Agent
+  This provides a driver and a monitor for the :ref:`Ibex Memory Interface Protocol<lsu-protocol>`.
+  The driver provides fully randomised and configurable timings for responses and randomisation of error responses.
+  Two agents are instaniated, one for the data memory interface the other for the instruction memory interface.
+  Read data for memory responses is provided from a backing memory, writes requests update the contents of the backing memory.
```

Comment:
Nit: writes requests -> write requests

Reply:
Also, probably a semicolon between the two halves.